cli/azd/cmd/provision.go

```diff
@@ -166,16 +166,25 @@ func (p *provisionAction) Run(ctx context.Context) (*actions.ActionResult, error
 	if err != nil {
 		return nil, fmt.Errorf("creating provisioning manager: %w", err)
 	}
-
-	deploymentPlan, err := infraManager.Plan(ctx)
-	if err != nil {
-		return nil, fmt.Errorf("planning deployment: %w", err)
-	}
+	var deployResult *provisioning.DeployResult
 
 	provisioningScope := infra.NewSubscriptionScope(
 		p.azCli, p.env.GetLocation(), p.env.GetSubscriptionId(), p.env.GetEnvName(),
 	)
-	deployResult, err := infraManager.Deploy(ctx, deploymentPlan, provisioningScope)
+	projectEventArgs := project.ProjectLifecycleEventArgs{
+		Project: p.projectConfig,
+	}
+
+	err = p.projectConfig.Invoke(ctx, project.ProjectEventProvision, projectEventArgs, func() error {
```

Comment:
Not immediately clear: do we end up firing `preprovision` twice when you run `provision`?

Reply:
What I understand from this PR is that we are enabling registering multiple handlers for the same hook (which was an error in the past). 
This enables internal and external hooks, where a handler is added for external hook (in case customer set the hook) and another handler can be added internally by azd to create a hook for azd.

Then, we are calling `addPreProvisionChecks()` to basically register the function as a `preprovision hook` (internal).  That's what I understand from the code.
My questions are:
- Why does it need to be an internal hook.
- It looks like project (service_target_containerapp.go) is taking a decision to do some stuff when the project is `provisioned` by hooking into `preprovision`.  This looks like a complicated bridge/workaround to me, because a project now knows about something that it is not part of its interface (ServiceTarget interface) and it is depending on it.
- Can we update `ServiceTarget interface` and add a new method like `provision()` .  Then make command `provision` to run `provision()` for every service as part of the main `provision` flow?  .   This would feel more general approach and would set a patter for any `ServiceTarget implementation` to define things/validation/extra-steps for the service type as part of its interface and not as a bridge using `azd hooks`